### PR TITLE
Improve mango notebook

### PR DESCRIPTION
**Pull request by bioinfornatics**
This update includes:
- compatibility with hortonworks (SPARK_HOME lookup)
- remove the useless use of ls, echo, fork and so on (i.e pure bash)
- find correctly the full path of the requested program
- use `PYSPARK_PYTHON`, `PYSPARK_DRIVER_PYTHON`, `PYSPARK_DRIVER_PYTHON_OPTS`, `PYSPARK_SUBMIT_ARGS`  environnement variable 
- Add error handling

### CODE REVIEW

**Comment by akmorrow13** [comment]
Thanks for the contribution @bioinfornatics! I made PR https://github.com/bigdatagenomics/mango/pull/430 which adds a Jenkins test for the notebook. Once that is merged I will trigger this test.

**Comment by bioinfornatics** [comment]
Well done @akmorrow13. Always use tests 😉.

**Comment by akmorrow13** [comment]
Jenkins, test this please.

**Comment by bioinfornatics** [comment]
@akmorrow13 current error is:
`Exception in thread "main" java.lang.IllegalArgumentException: pyspark does not support any application `

is it possible to run `mango-notebook` with debug output  ?
In order to do this at line [231](https://github.com/bigdatagenomics/mango/blob/master/scripts/jenkins-test#L231) prefix command line by `bash -xv`

```bash
bash -xv ${PROJECT_ROOT}/bin/mango-notebook  &> ${MANGO_NOTEBOOK_LOGS} &
```

Indeed I need to see the trace to fix it.

Thanks a lot

**Comment by akmorrow13** [comment]
@bioinfornatics yes you can add `bash -xv` to [scripts/jenkins-test](https://github.com/bigdatagenomics/mango/blob/master/scripts/jenkins-test#L231) in this PR to see the output. Right now, the [pyspark args](https://github.com/bigdatagenomics/mango/pull/429/files#diff-a58429e700d73630640786935d81460cR139) are throwing the error.

**Comment by bioinfornatics** [comment]
So I rebase on master and put my commit at the end to have a clean history. I do not if that follow your git flow.

**Comment by bioinfornatics** [comment]
I will try another way in order to keep bubble merge in the history

**Comment by bioinfornatics** [comment]
done 
Is it possible to run the test in order to see the stack trace, please?

**Comment by bioinfornatics** [comment]
Jenkins, test this please.

**Comment by bioinfornatics** [comment]
@akmorrow13 could you launch the test for debug purpose please ?

**Comment by akmorrow13** [comment]
Jenkins, add to whitelist.

**Comment by akmorrow13** [comment]
Jenkins, test this please.

**Comment by bioinfornatics** [comment]
Jenkins, test this please.

File: bin/mango-notebook [hunk 1]
```diff
@@ -47,29 +86,55 @@ fi
...
+  PY4J_ZIP_ALL_VERSIONS=( ${SPARK_HOME}/python/lib/py4j-+([0-9]).+([0-9]).+([0-9])-src.zip )
+  # use the latest version
+  PY4J_ZIP="${PY4J_ZIP_ALL_VERSIONS[-1]}"
+  if [[ ! -e ${PY4J_ZIP} ]]; then
+    print_err "${PY4J_ZIP} is not found; Aborting."
+    exit 1
+  fi
+fi
 
-PY4J_ZIP="$(ls -1 "${SPARK_HOME}/python/lib" | grep py4j)"
```
Comment (akmorrow13) [inline bin/mango-notebook]: We need access to this zip to run correctly on workers. Is there a reason it is removed?

File: bin/mango-notebook [hunk 2]
```diff
@@ -47,29 +86,55 @@ fi
...
+fi
+
+if [[ -z ${PYSPARK_DRIVER_PYTHON_OPTS} ]]; then
+  export PYSPARK_DRIVER_PYTHON_OPTS="notebook ${NOTEBOOK_ARGS}"
+else
+  export PYSPARK_DRIVER_PYTHON_OPTS="${PYSPARK_DRIVER_PYTHON_OPTS} ${NOTEBOOK_ARGS}"
+fi
+
+if [[ -d "${SPARK_HOME}/python/lib" ]]; then
+  PY4J_ZIP_ALL_VERSIONS=( ${SPARK_HOME}/python/lib/py4j-+([0-9]).+([0-9]).+([0-9])-src.zip )
```
Comment (akmorrow13) [inline bin/mango-notebook]: Generates error:
```
./bin/mango-notebook: line 126: PY4J_ZIP_ALL_VERSIONS: bad array subscript
ERROR:  is not found; Aborting.
```

File: bin/mango-notebook [hunk 3]
```diff
@@ -47,29 +86,55 @@ fi
...
-PY4J_ZIP="$(ls -1 "${SPARK_HOME}/python/lib" | grep py4j)"
 export PYTHONPATH=${SPARK_HOME}/python:${SPARK_HOME}/python/lib/${PY4J_ZIP}:${PYTHONPATH}
 
 ${PYSPARK} \
   --conf spark.serializer=org.apache.spark.serializer.KryoSerializer \
   --conf spark.kryo.registrator=org.bdgenomics.mango.serialization.MangoKryoRegistrator \
   --jars ${MANGO_CLI_JAR} \
   --driver-class-path ${MANGO_CLI_JAR} \
-  "${SPARK_ARGS[@]}"
+  ${SPARK_ARGS[@]}
```
Comment (akmorrow13) [inline bin/mango-notebook]: Why are the quotes removed?

File: bin/mango-notebook (hunk 2 above)
Comment (bioinfornatics) [reply bin/mango-notebook]: Hi @akmorrow13 thanks for the review. 
Indeed I did a mistake with the regex
current:
```
{SPARK_HOME}/python/lib/py4j-+([0-9]).+([0-9]).+([0-9])-src.zip
```
should be:
```
${SPARK_HOME}/python/lib/py4j-([0-9]+)\.([0-9]+)\.([0-9]+)-src.zip
```

The bad array subscript show that PY4J_ZIP_ALL_VERSIONS is empty

File: bin/mango-notebook (hunk 3 above)
Comment (bioinfornatics) [reply bin/mango-notebook]: I thought it was better to expand parameters.
I show below 2 use case tell me which is expected, maybe I did a mistake.
before to start a look at the test function:
```bash
test_param(){  
  echo "number of provided parameter: $#"
  echo "parameter 1: $1"
  echo "value 1: $2"
  echo "parameter 2: $3"
  echo "value 2: $4"
}
```

In our case `SPARK_ARGS` is an array defined as:
```bash
 SPARK_ARGS=( "--master yarn" "--deploy-mode cluster" )
```

- **case 1** call `test_param` using quote while expanding `SPARK_ARGS` array
```bash
test_param  "${SPARK_ARGS[@]}"
number of provided parameter: 2
parameter 1: --master yarn
value 1: --deploy-mode cluster
parameter 2:
value 2:
```
- **case 2** call `test_param` without quote while expanding `SPARK_ARGS` array
```bash
test_param  ${SPARK_ARGS[@]}
number of provided parameter: 4
parameter 1: --master
value 1: yarn
parameter 2: --deploy-mode
value 2: cluster
```

What is the expected behavior?

thanks @akmorrow13

File: bin/mango-notebook (hunk 2 above)
Comment (bioinfornatics) [reply bin/mango-notebook]: So the jenkins test script should maybe be updated accordingly to catch your error, no ?

Comment (bioinfornatics) [reply bin/mango-notebook]: @akmorrow13 did you have still the problem with the latest update ?

File: bin/mango-notebook [hunk 4]
```diff
@@ -17,59 +17,149 @@
...
+# Retrieve the path of py4j zip archive used by current spark
+get_py4jzip(){
+  local -a py4j_zip_all_versions
+  local py4j_zip=""
+  if [[ -d "${SPARK_HOME}/python/lib" ]]; then
+    py4j_zip_all_versions=( ${SPARK_HOME}/python/lib/py4j-+([0-9])\.+([0-9])?(\.+([0-9]))-src.zip )
+    # use the latest version
+    py4j_zip="${py4j_zip_all_versions[-1]}"
+    if [[ ! -e ${py4j_zip} ]]; then
+      err_report ${LINENO} $? "${py4j_zip} is not found; Aborting."
```
Comment (akmorrow13) [inline bin/mango-notebook]: When this error is found, it only exits out of the function, but the script succeeds. This failure should end the script.

File: bin/mango-notebook [hunk 5]
```diff
@@ -17,59 +17,149 @@
...
-  NOTEBOOK_ARGS=("${PRE_DD[@]}")
-fi
+# Retrieve the path of py4j zip archive used by current spark
+get_py4jzip(){
+  local -a py4j_zip_all_versions
+  local py4j_zip=""
+  if [[ -d "${SPARK_HOME}/python/lib" ]]; then
+    py4j_zip_all_versions=( ${SPARK_HOME}/python/lib/py4j-+([0-9])\.+([0-9])?(\.+([0-9]))-src.zip )
+    # use the latest version
+    py4j_zip="${py4j_zip_all_versions[-1]}"
```
Comment (akmorrow13) [inline bin/mango-notebook]: This indexing does not work on mac OS. We need to get the array length:
`"${py4j_zip_all_versions[${#py4j_zip_all_versions[@]}-1]}"`

**Comment by bioinfornatics** [comment]
up

**Comment by akmorrow13** [comment]
lgtm!

**Comment by akmorrow13** [comment]
Thanks @bioinfornatics!

**Pull request merged**